eudemo/eudemo/ivc/panelling/_designer.py

```diff
@@ -0,0 +1,203 @@
+            boundary, opt_problem, max_retries
+        )
+
+        # Make sure we warn about broken tolerances this time.
+        if num_retries == max_retries and (
+            x_opt is None or not opt_problem.opt.check_constraints(x_opt, warn=True)
+        ):
+            bluemira_warn(
+                "Could not solve panelling optimisation problem: no feasible "
+                "solution found. Try reducing the minimum length and/or increasing "
+                "the maximum allowed angle."
+            )
+        if x_opt is None or opt_problem.constraints_violated_by(x_opt, 1):
+            return initial_solution
+
+        return opt_problem.paneller.joints(x_opt)
+
+    def mock(self) -> np.ndarray:
+        """
+        Mock the design problem, returning the initial guess for panel placement.
+
+        This guarantees that panels will always fully contain the given
+        boundary, but does not guarantee the maximum angle and minimum
+        length constraints are honoured.
+        """
+        boundary = self.wall_boundary.discretize(byedges=True).xyz[[0, 2], :]
+        paneller = Paneller(
+            boundary, self.params.fw_a_max.value, self.params.fw_dL_min.value
+        )
+        return paneller.joints(paneller.x0)
+
+    def _run_opt_problem_with_retries(
+        self,
+        boundary: np.ndarray,
+        opt_problem: PanellingOptProblem,
+        max_retries: int,
+    ) -> Tuple[Union[np.ndarray, None], PanellingOptProblem, int]:
+        """
+        Run the minimise panel length optimisation problem.
+
+        If the optimisation fails, retry with an extra panel
+        ``max_retries`` times. Return ``None`` if the optimiser crashes,
+        which it can often do, with 'more than iter SQP iterations'.
+        """
+        try:
+            x_opt = opt_problem.optimise()
+        except ExternalOptError:
+            # Avoid 'more than iter SQP iterations' errors stopping the
+            # design.
+            # Ignoring the error here is OK, as the optimiser prints a
+            # warning and we either try again with more panels, or
+            # return our initial guess as a fall back.
+            x_opt = None
+        iter_num = 0
+        while (
+            x_opt is None or not opt_problem.opt.check_constraints(x_opt, warn=False)
+        ) and iter_num < max_retries:
+            # We couldn't satisfy the constraints on our last attempt,
+            # so try increasing the number of panels.
+            # Note we're actually increasing the number of panels by 1
+            # by adding 3 below, as there are two more panels than
+            # optimisation parameters.
+            n_panels = opt_problem.n_opts + 3
+            opt_problem = self._set_up_opt_problem(boundary, n_panels)
+            try:
+                x_opt = opt_problem.optimise(check_constraints=False)
+            except ExternalOptError:
+                x_opt = None
+            iter_num += 1
```

Comment:
I did explore that a little. The optimum will always end up being the one with most panels though, as that solution will inevitably hug the boundary more, and have the smallest length. So we may as well just run the optimisation on `2*len(initial_joints)` to begin with. 

The way it's currently set up (which is in the spirit of how the problem appears to be approached in BLUEPRINT), we're getting the minimum number of panels for which the constraints are satisfied. Doing the mixed optimisation problem you're describing (with no upper bound) would result in the maximum number of panels for which the constraints are satisfied (as this would inevitably be the minimum length). Is that what we actually want?

Reply:
No, you are right. Generally we want few panels. But there's probably a pareto front with a second objective there we are ignoring, which is fine.